### PR TITLE
A bag of small (but useful) fixes

### PR DESCRIPTION
**Pull request by danieldk**
## Description

* We were still passing around `eos_id`, which is unused since we have an eos stopping condition.
* Do not assume inference in the generator (remove hardcoded `model.eval()` and `torch.no_grad()`. Provide `eval`/`train` methods on the generator instead.
* Fix an `AutoGenerator` test.
* Construct attention masks with the same type as query/key to avoid accidental type widening and unbreak Torch SDP in non-float32.

### Types of change

Bugfix.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/generation/generator.py [hunk 1]
```diff
@@ -101,14 +97,13 @@ def generate(
...
         while True:
-            with torch.no_grad():
-                output = self.model(
-                    state.last_step_ids,
-                    attention_mask=AttentionMask(state.attention_mask),
-                    cache=state.cache,
-                    store_cache=True,
-                    positions=state.positions,
-                )
+            output = self.model(
```
[the remark this answers is not in the document]
Comment (shadeMe) [reply curated_transformers/generation/generator.py:100]: Right. Then it's probably not a bad idea to mention in the docstring of `train` and `eval` that that the user is responsible for disabling gradients during inference/eval.

[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/generation/generator.py]: Reverted this particular change after more discussion in Slack.
